### ISSUE TITLE
Errors on nomos regex empty subexpressions

### ISSUE THREAD

**Issue by AMDmi3** [fossology/fossology#822]
Already reported in #754 and not fixed there. nomos (standalone build, fossology 3.1.0, FreeBSD 11.0 amd64, pcre 8.40) fails to compile regexes containing empty subexpressions. Steps to reproduce:

```
% echo 'Eclipse Distribution License' > /tmp/test                                                                                                                          
% nomossa /tmp/test                                                                                                                                                        
Compile failed, regex #1369                                                                                                                                                                                       
regex = "Eclipse Distribution License[ -](v |v|)1\.?0"                                                                                                                                                            
FATAL nomos_regex.c.49: regcomp failure: empty (sub)expression                      
```

```
% echo 'program is linked with microsoft distributable code' > /tmp/test                                                                                                   
% nomossa /tmp/test                                                                                                                                                        
Compile failed, regex #711                                                                                                                                                                                        
regex = "program is linked with (|and uses) microsoft distributable code"                                                                                                                                         
FATAL nomos_regex.c.49: regcomp failure: empty (sub)expression                                                                                                                                                    
```

**Comment by mcjaeger** [comment]
I think your platform is kind of exotic to the Fossology community base (although freebsd is a good OS!), however, I am not sure if anyone else uses freebsd, so compilation issues like this are hard to solve. 

How to install fossology on Freebsd actually?

**Comment by AMDmi3** [comment]
> I think your platform is kind of exotic to the Fossology community base (although freebsd is a good OS!), however, I am not sure if anyone else uses freebsd, so compilation issues like this are hard to solve.

This gave me a hint on that this is platform specific, and yes, nomos seem to use system POSIX regexp engine (I though it used pcre), which may behave differently on FreeBSD.

And yes, it does, this test program

```
#include <regex.h>
#include <stdio.h>

int main() {
    int res;
    char errbuf[1024];
    regex_t re;

    res = regcomp(&re, "foo(|bar)baz", REG_EXTENDED);
    if (res != 0) {
        regerror(res, &re, errbuf, sizeof(errbuf));
        fprintf(stderr, "regcomp failure: %s\n", errbuf);
        return 1;
    } else {
        fprintf(stderr, "regcomp ok\n");
    }

    return 0;
}
```

will print `regcomp failure: empty (sub)expression` on FreeBSD, though it would print `regcomp ok` on Linux I have by hand.

Now, let's look at [the standard](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_04_03) to determine who's right:

```
|
    The <vertical-line> is special except when used in a bracket
    expression (see RE Bracket Expression). A <vertical-line>
    appearing first or last in an ERE, or immediately following a
    <vertical-line> or a <left-parenthesis>, or immediately preceding
    a <right-parenthesis>, produces undefined results.
```

So, regardless on the OS the code is run on, regexps used in nomos are illegal. It looks like empty subexpressions could be rather easily avoided by using constructs like `(foo|bar)?`

> How to install fossology on Freebsd actually?

There's port specifically for standalone nomos scanner, [devel/fossology-nomos-standalone](http://www.freshports.org/devel/fossology-nomos-standalone)

**Comment by AMDmi3** [comment]
It turns out that the problem is not as severe as I though it is without checking first. There are only 3 problematic RE's, all are fixed with #879
